runtime/ops/ops_buffer.rs

```diff
@@ -0,0 +1,281 @@
+use std::future::Future;
+use std::rc::Rc;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct RequestHeader {
+  pub promise_id: u32,
+  pub argument: u32,
+}
+
+impl RequestHeader {
+  pub fn from_raw(bytes: &[u8]) -> Option<Self> {
+    if bytes.len() < 2 * 4 {
+      return None;
+    }
+    let ptr = bytes.as_ptr() as *const u32;
+    let ints = unsafe { std::slice::from_raw_parts(ptr, 2) };
+    Some(Self {
+      promise_id: ints[0],
+      argument: ints[1],
+    })
+  }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct ResponseHeader {
+  pub promise_id: u32,
+  pub status: u32,
+  pub result: u32,
+}
+
+impl Into<[u8; 12]> for ResponseHeader {
+  fn into(self) -> [u8; 12] {
+    let slice = [self.promise_id, self.status, self.result];
+    unsafe { std::mem::transmute::<[u32; 3], [u8; 3 * 4]>(slice) }
+  }
+}
+
+pub trait ValueOrVector {
+  fn value(&self) -> u32;
+  fn vector(self) -> Option<Vec<u8>>;
+}
+
+impl ValueOrVector for Vec<u8> {
+  fn value(&self) -> u32 {
+    self.len() as u32
+  }
+  fn vector(self) -> Option<Vec<u8>> {
+    Some(self)
+  }
+}
+
+impl ValueOrVector for u32 {
+  fn value(&self) -> u32 {
+    *self
+  }
+  fn vector(self) -> Option<Vec<u8>> {
+    None
+  }
+}
+
+/// Creates an op that passes data synchronously using raw ui8 buffer.
+///
+/// The provided function `op_fn` has the following parameters:
+/// * `&mut OpState`: the op state, can be used to read/write resources in the runtime from an op.
+/// * `argument`: the i32 value that is passed to the Rust function.
+/// * `&mut [ZeroCopyBuf]`: raw bytes passed along.
+///
+/// `op_fn` returns an array buffer value, which is directly returned to JavaScript.
+///
+/// When registering an op like this...
+/// ```ignore
+/// let mut runtime = JsRuntime::new(...);
+/// runtime.register_op("hello", deno_core::buffer_op_sync(Self::hello_op));
+/// ```
+///
+/// ...it can be invoked from JS using the provided name, for example:
+/// ```js
+/// Deno.core.ops();
+/// let result = Deno.core.bufferOpSync("function_name", args);
+/// ```
+///
+/// The `Deno.core.ops()` statement is needed once before any op calls, for initialization.
+/// A more complete example is available in the examples directory.
+pub fn buffer_op_sync<F, R>(op_fn: F) -> Box<OpFn>
+where
+  F: Fn(&mut OpState, u32, &mut [ZeroCopyBuf]) -> Result<R, AnyError> + 'static,
+  R: ValueOrVector,
+{
+  Box::new(move |state: Rc<RefCell<OpState>>, bufs: BufVec| -> Op {
+    let mut bufs_iter = bufs.into_iter();
+    let record_buf = bufs_iter.next().expect("Expected record at position 0");
+    let mut zero_copy = bufs_iter.collect::<BufVec>();
+
+    let req_header = match RequestHeader::from_raw(&record_buf) {
+      Some(r) => r,
+      None => {
+        let error_class = b"TypeError";
+        let error_message = b"Unparsable control buffer";
+        let len = error_class.len() + error_message.len();
+        let padding = &[b' ', b' ', b' '][0..(4 - (len & 3)) & 3];
```

Comment:
This padding logic is repeated several times. It would be good to abstract it out into a standalone helper function with its own test to prod the edge cases.

Reply:
Ok I created `gen_padding_32bit` for this.